rust/theoros/src/storage/checkpoints.rs

```diff
@@ -0,0 +1,83 @@
+use std::collections::{BTreeMap, HashMap};
+use std::sync::Arc;
+
+use starknet::core::types::Felt;
+use tokio::sync::RwLock;
+
+use crate::types::hyperlane::{DispatchEvent, SignedCheckpointWithMessageId};
+
+/// Mapping between messages nonces and their corresponding Event.
+#[derive(Clone, Default)]
+pub struct UnsignedCheckpointsStorage(Arc<RwLock<BTreeMap<u32, DispatchEvent>>>);
+
+impl UnsignedCheckpointsStorage {
+    /// Insert a new mapping between a nonce & an Event.
+    pub async fn add(&self, nonce: u32, event: &DispatchEvent) {
+        let mut lock = self.0.write().await;
+        lock.insert(nonce, event.clone());
+    }
+
+    /// Retrieve all nonces currently stored, in sorted order.
+    pub async fn nonces(&self) -> Vec<u32> {
+        let lock = self.0.read().await;
+        lock.keys().cloned().collect()
+    }
+
+    /// Remove a nonce from the storage.
+    pub async fn remove(&self, nonce: u32) {
+        let mut lock = self.0.write().await;
+        lock.remove(&nonce);
+    }
+
+    /// Get the event associated with a nonce.
+    pub async fn get(&self, nonce: u32) -> Option<DispatchEvent> {
+        let lock = self.0.read().await;
+        lock.get(&nonce).cloned()
+    }
+}
+
+/// Mapping between the validators and their signed checkpoint for a given nonce.
+#[derive(Debug, Default)]
+pub struct SignedCheckpointsStorage(RwLock<HashMap<(Felt, u32), SignedCheckpointWithMessageId>>);
```

Comment:
we can look into https://github.com/xacrimon/dashmap well maintained replacement for this data struct